File: 大三/人工智能导论/实验/实验二/AI实验二_37220232203780_马鑫/BP算法/BP算法.py
import sys
import math
import os
import numpy as np

# ============================================================
#  工具函数：矩阵 / 向量的动态分配与释放
# ============================================================

def matrix_allocate(n_rows, n_cols):
    """分配 n_rows x n_cols 的二维 double 数组"""
    return np.zeros((n_rows, n_cols), dtype=np.float64)

def vector_allocate(n_cols):
    """分配长度为 n_cols 的一维 double 数组"""
    return np.zeros(n_cols, dtype=np.float64)

def matrix_free(mat):
    """释放矩阵（Python 中由 GC 自动管理，此处仅作占位兼容）"""
    pass

# ============================================================
#  核心网络结构体
# ============================================================

class BPSavedParams:
    """保存每次运行时动态分配的网络层矩阵"""
    def __init__(self):
        self.out0 = None   # 输入层
        self.out1 = None   # 隐含层
        self.delta1 = None # 隐含层误差修正量
        self.delw1 = None  # 输入-隐含层权值变化量
        self.w1 = None     # 输入-隐含层权值
        self.out2 = None   # 输出层
        self.delta2 = None # 输出层误差修正量
        self.delw2 = None  # 隐含-输出层权值变化量
        self.w2 = None     # 隐含-输出层权值
        self.target = None # 目标输出值
        self.pattern_id = None  # 模式识别标志

# ============================================================
#  主程序
# ============================================================

def main():
    # ---------- 默认超参数 ----------
    eta = 0.15            # 学习速率
    alpha = 0.075         # 势态因子（动量项）
    n_report_errors = 100 # 误差报告频率
    error_level = 0.02    # 满意的误差水平
    monitor_error = False # 是否强制监控误差

    # ---------- 解析命令行可选参数 ----------
    args = sys.argv[1:]
    idx = 0
    while idx < len(args):
        arg = args[idx]
        if not arg.startswith('-'):
            break
        opt = arg[1]
        if opt == 'e' and idx + 1 < len(args):
            n_report_errors = int(args[idx + 1])
            idx += 2
        elif opt == 'd' and idx + 1 < len(args):
            error_level = float(args[idx + 1])
            idx += 2
        else:
            break

    # 必须提供运行配置文件名
    if idx >= len(args):
        print("Usage: python {} {-en -df} runfilename".format(sys.argv[0]))
        print("  -e n => report error every n iterations (default 100)")
        print("  -d f => done if mean squared error < f (default 0.02)")
        sys.exit(1)

    run_filename = args[idx]
    progname = sys.argv[0]

    # ---------- 打开运行配置文件 ----------
    if not os.path.exists(run_filename):
        print("{}: can't open file {}".format(progname, run_filename))
        sys.exit(1)

    with open(run_filename, 'r') as fp_run:
        # 读第一行：运行次数
        n_runs = int(fp_run.readline().strip())

        # ---------- 每次运行 ----------
        for r in range(n_runs):
            # 读第二行：所有文件路径 + 网络参数
            tokens = fp_run.readline().strip().split()
            sz_results   = tokens[0]
            sz_error     = tokens[1]
            sz_pattern   = tokens[2]
            sz_weights   = tokens[3]
            sz_weights_out = tokens[4]
            n_patterns     = int(tokens[5])
            n_iterations   = int(tokens[6])
            n_input_nodes  = int(tokens[7])
            n_hidden_nodes = int(tokens[8])
            n_output_nodes = int(tokens[9])
            eta            = float(tokens[10])
            alpha          = float(tokens[11])

            net = BPSavedParams()

            # ---------- 动态存储分配 ----------
            net.out0   = matrix_allocate(n_patterns, n_input_nodes)
            net.out1   = matrix_allocate(n_patterns, n_hidden_nodes)
            net.out2   = matrix_allocate(n_patterns, n_output_nodes)
            net.delta2 = matrix_allocate(n_patterns, n_output_nodes)
            net.delw2  = matrix_allocate(n_output_nodes, n_hidden_nodes + 1)  # 含偏置列
            net.w2     = matrix_allocate(n_output_nodes, n_hidden_nodes + 1)
            net.delta1 = matrix_allocate(n_patterns, n_hidden_nodes)
            net.delw1  = matrix_allocate(n_hidden_nodes, n_input_nodes + 1)   # 含偏置列
            net.w1     = matrix_allocate(n_hidden_nodes, n_input_nodes + 1)
            net.target = matrix_allocate(n_patterns, n_output_nodes)
            net.pattern_id = vector_allocate(n_patterns)

            # ---------- 读初始权矩阵 ----------
            if not os.path.exists(sz_weights):
                print("{}: can't open file {}".format(progname, sz_weights))
                sys.exit(1)

            with open(sz_weights, 'r') as fp_w:
                all_vals = fp_w.read().split()
                ptr = 0
                # 输入-隐含层权值: nHiddenNodes x (nInputNodes+1)
                for h in range(n_hidden_nodes):
                    for i in range(n_input_nodes + 1):
                        net.w1[h][i] = float(all_vals[ptr])
                        net.delw1[h][i] = 0.0
                        ptr += 1
                # 隐含-输出层权值: nOutputNodes x (nHiddenNodes+1)
                for j in range(n_output_nodes):
                    for h in range(n_hidden_nodes + 1):
                        net.w2[j][h] = float(all_vals[ptr])
                        net.delw2[j][h] = 0.0
                        ptr += 1

            # ---------- 读所有训练模式 ----------
            if not os.path.exists(sz_pattern):
                print("{}: can't open file {}".format(progname, sz_pattern))
                sys.exit(1)

            p = 0
            with open(sz_pattern, 'r') as fp_p:
                all_lines = fp_p.read().split()
                val_ptr = 0
                while p < n_patterns:
                    # 检查是否有足够的数据
                    needed = n_input_nodes + n_output_nodes + 1  # 输入 + 目标 + ID
                    if val_ptr + needed > len(all_lines):
                        break
                    # 读输入层
                    for i in range(n_input_nodes):
                        net.out0[p][i] = float(all_lines[val_ptr])
                        val_ptr += 1
                    # 读目标输出
                    for j in range(n_output_nodes):
                        net.target[p][j] = float(all_lines[val_ptr])
                        val_ptr += 1
                    # 读模式识别标志
                    net.pattern_id[p] = float(all_lines[val_ptr])
                    val_ptr += 1
                    p += 1

            if p < n_patterns:
                print("{}: {} out of {} patterns read".format(progname, p, n_patterns))
                n_patterns = p

            # ---------- 打开误差输出文件 ----------
            try:
                fp_error = open(sz_error, 'w')
            except IOError:
                print("{}: can't open file {}".format(progname, sz_error))
                sys.exit(1)

            if n_iterations > 1:
                print("Training...")
            else:
                print("Testing...")

            # ==================== 开始迭代循环 ====================
            error = 0.0
            q = 0
            for q in range(n_iterations):
                for p in range(n_patterns):
                    # ---- 隐含层前向传播 ----
                    for h in range(n_hidden_nodes):
                        s = net.w1[h][n_input_nodes]  # 偏置项
                        for i in range(n_input_nodes):
                            s += net.w1[h][i] * net.out0[p][i]
                        net.out1[p][h] = 1.0 / (1.0 + math.exp(-s))

                    # ---- 输出层前向传播 ----
                    for j in range(n_output_nodes):
                        s = net.w2[j][n_hidden_nodes]  # 偏置项
                        for h in range(n_hidden_nodes):
                            s += net.w2[j][h] * net.out1[p][h]
                        net.out2[p][j] = 1.0 / (1.0 + math.exp(-s))

                    # ---- 输出层误差修正量 ----
                    for j in range(n_output_nodes):
                        net.delta2[p][j] = (net.target[p][j] - net.out2[p][j]) \
                                           * net.out2[p][j] * (1.0 - net.out2[p][j])

                    # ---- 隐含层误差修正量 ----
                    for h in range(n_hidden_nodes):
                        s = 0.0
                        for j in range(n_output_nodes):
                            s += net.delta2[p][j] * net.w2[j][h]
                        net.delta1[p][h] = s * net.out1[p][h] * (1.0 - net.out1[p][h])

                # ---- 调整隐含-输出层的权值（批量累加后更新）----
                for j in range(n_output_nodes):
                    # 偏置项
                    s = 0.0
                    for p in range(n_patterns):
                        s += net.delta2[p][j]
                    dw = eta * s + alpha * net.delw2[j][n_hidden_nodes]
                    net.w2[j][n_hidden_nodes] += dw
                    net.delw2[j][n_hidden_nodes] = dw

                    # 权值项
                    for h in range(n_hidden_nodes):
                        s = 0.0
                        for p in range(n_patterns):
                            s += net.delta2[p][j] * net.out1[p][h]
                        dw = eta * s + alpha * net.delw2[j][h]
                        net.w2[j][h] += dw
                        net.delw2[j][h] = dw

                # ---- 调整输入-隐含层的权值（批量累加后更新）----
                for h in range(n_hidden_nodes):
                    # 偏置项
                    s = 0.0
                    for p in range(n_patterns):
                        s += net.delta1[p][h]
                    dw = eta * s + alpha * net.delw1[h][n_input_nodes]
                    net.w1[h][n_input_nodes] += dw
                    net.delw1[h][n_input_nodes] = dw

                    # 权值项
                    for i in range(n_input_nodes):
                        s = 0.0
                        for p in range(n_patterns):
                            s += net.delta1[p][h] * net.out0[p][i]
                        dw = eta * s + alpha * net.delw1[h][i]
                        net.w1[h][i] += dw
                        net.delw1[h][i] = dw

                # ---- 计算均方根误差 ----
                if monitor_error or (q % n_report_errors == 0):
                    error = 0.0
                    for p in range(n_patterns):
                        for j in range(n_output_nodes):
                            temp = net.target[p][j] - net.out2[p][j]
                            error += temp * temp
                    error /= (n_patterns * n_output_nodes)
                    print("\riteration {}-{} Error {:f}".format(q, n_iterations, error), end="")
                    monitor_error = False

                    if q % n_report_errors == 0:
                        fp_error.write("{} {:f}\n".format(q, error))

                    if error < error_level:
                        break

            # ==================== 迭代循环结束 ====================

            # 计算最终平均误差
            error = 0.0
            for p in range(n_patterns):
                for j in range(n_output_nodes):
                    temp = net.target[p][j] - net.out2[p][j]
                    error += temp * temp
            error /= (n_patterns * n_output_nodes)

            print("\nIteration {}-{} Error {:f}".format(q, n_iterations, error))
            fp_error.write("\n{} {:f}\n".format(q, error))
            fp_error.close()

            # ---------- 输出最终权值文件 ----------
            try:
                fp_wout = open(sz_weights_out, 'w')
            except IOError:
                print("{}: can't write file {}".format(progname, sz_weights_out))
                sys.exit(1)

            items_per_line = 8
            # 输入-隐含层权值
            count = 0
            for h in range(n_hidden_nodes):
                for i in range(n_input_nodes + 1):
                    fp_wout.write("{:g}".format(net.w1[h][i]))
                    count += 1
                    if count % items_per_line == 0:
                        fp_wout.write('\n')
                    else:
                        fp_wout.write(' ')
            # 隐含-输出层权值
            count = 0
            for j in range(n_output_nodes):
                for h in range(n_hidden_nodes + 1):
                    fp_wout.write("{:g}".format(net.w2[j][h]))
                    count += 1
                    if count % items_per_line == 0:
                        fp_wout.write('\n')
                    else:
                        fp_wout.write(' ')
            fp_wout.close()

            # ---------- 输出结果文件 ----------
            try:
                fp_res = open(sz_results, 'w')
            except IOError:
                print("{}: can't write file {}".format(progname, sz_results))
                fp_res = sys.stderr

            for p in range(n_patterns):
                line = "{} ".format(p)
                for j in range(n_output_nodes):
                    line += "{:f} ".format(net.out2[p][j])
                line += " {:>-6.0f}\n".format(net.pattern_id[p])
                fp_res.write(line)
            fp_res.close()

            # ---------- 释放存储（Python 自动回收，此处仅作示意）----------
            matrix_free(net.out0)
            matrix_free(net.out1)
            matrix_free(net.delta1)
            matrix_free(net.delw1)
            matrix_free(net.w1)
            matrix_free(net.out2)
            matrix_free(net.delta2)
            matrix_free(net.delw2)
            matrix_free(net.w2)
            matrix_free(net.target)
            del net.pattern_id


if __name__ == '__main__':
    main()

File: 大三/人工智能导论/实验/实验二/AI实验二_37220232203780_马鑫/BP算法/BP算法.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <fstream>
#include <iostream>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <conio.h>
#include <cstring>

using namespace std;

#define ESC 27 
#define ERRORLEVEL 0.02 
#define ITEMS 8 

//指针的动态存储 
typedef double* PFLOAT;
typedef PFLOAT VECTOR;
typedef PFLOAT* MATRIX;

void VectorAllocate(VECTOR* vector, int nCols);
void AllocateCols(PFLOAT matrix[], int nRows, int nCols);
void MatrixAllocate(MATRIX* pmatrix, int nRows, int nCols);
void MatrixFree(MATRIX matrix, int nRows);

//定义网络层的存储 
//输入层,输出层及权值等的指针; 
MATRIX out0;      //输入层 
MATRIX out1;      //隐含层 
MATRIX delta1;    //隐含层中的误差修正量 
MATRIX delw1;     //输入与隐含层之间权值的变化 
MATRIX w1;        //输入与隐含层之间的权值 
MATRIX out2;      //输出层; 
MATRIX delta2;    //输出层中的误差修正量 
MATRIX delw2;     //隐含与输出层之间权值的变化 
MATRIX w2;        //隐含与输出层间的权值 
MATRIX target;    //目标输出值 
VECTOR PatternID; //每一个被储存模式的识别标志 

int main(int argc, char* argv[]) // 修复：VS要求main返回值为int
{
    double eta = 0.15;        //默认的学习速率 
    double alpha = 0.075;     //默认的势态因子 
    int nReportErrors = 100;  //误差报告频率 
    double ErrorLevel = ERRORLEVEL; //满意的误差水平 
    char MonitorError = 0;    //表示为真 
    double error;             //最新的平方根误差之和 
    register int h;           //代表隐含层的值 
    register int i;           //代表输入层的值 
    register int j;           //代表输出层的值 
    int p,                    //模式数 
        q,                    // 
        r,                    //运行数 
        nPatterns,            //模式个数 
        nInputNodes,          //输入节点个数 
        nHiddenNodes,         //隐含层节点个数 
        nOutputNodes,         //输出层节点个数 
        nIterations,          // 
        nRuns;                //运行个数 

    FILE* fpRun,          //运行文件 
        * fpPattern,      //模式输入文件 
        * fpWeights,      //初始权值文件 
        * fpWeightsOut,   //最合权输出文件 
        * fpResults,      //结果输出文件 
        * fpError;        //误差输出文件 

    char szResults[66];   //变量文件 
    char szError[66];
    char szPattern[66];
    char szWeights[66];
    char szWeightsOut[66];
    char* progname = *argv;

    //读优化参数 
    for (; argc > 1; argc--)
    {
        char* arg = *++argv;
        if (*arg != '-') break;
        switch (*++arg)
        {
        case 'e': sscanf(++arg, "%d", &nReportErrors); break;
        case 'd': sscanf(++arg, "%lf", &ErrorLevel); break;
        default: break;
        }
    }

    if (argc < 2)
    {
        fprintf(stderr, "Usage:%s{-en,-df} runfilename\n", progname);
        fprintf(stderr, "-en=>report error every n iterations\n");
        fprintf(stderr, "-df=>done if mean squared error<f\n");
        system("pause");
        exit(1);
    }

    //打开文件读数据 
    if ((fpRun = fopen(*argv, "r")) == NULL)
    {
        fprintf(stderr, "%s:can't open file %s\n", progname, *argv);
        system("pause");
        exit(1);
    }

    //读第一行 
    fscanf(fpRun, "%d", &nRuns);

    //循环开始 
    for (r = 0; r < nRuns; r++)
    {
        // 修复：double类型必须使用%lf读取
        fscanf(fpRun, "%s %s %s %s %s %d %d %d %d %lf %lf",
            szResults,     //结果输出文件 
            szError,       //误差输出文件 
            szPattern,     //模式输入文件 
            szWeights,     //初始权值文件 
            szWeightsOut,  //最后权输出文件 
            &nPatterns,    //待学习的模式的个数 
            &nIterations,  //迭代次数 
            &nInputNodes,  //输入节点个数 
            &nHiddenNodes, //隐含节点个数 
            &nOutputNodes, //输出节点个数 
            &eta,          //学习速率 
            &alpha);       //动量因子 

        //所有数据的动态存储 
        MatrixAllocate(&out0, nPatterns, nInputNodes);
        MatrixAllocate(&out1, nPatterns, nHiddenNodes);
        MatrixAllocate(&out2, nPatterns, nOutputNodes);
        MatrixAllocate(&delta2, nPatterns, nOutputNodes);
        MatrixAllocate(&delw2, nOutputNodes, nHiddenNodes + 1); // 修复：原代码分配为nHiddenNodes，但后续访问了nHiddenNodes(偏置项)，会导致越界崩溃
        MatrixAllocate(&w2, nOutputNodes, nHiddenNodes + 1);
        MatrixAllocate(&delta1, nPatterns, nHiddenNodes);
        MatrixAllocate(&delw1, nHiddenNodes, nInputNodes + 1);
        MatrixAllocate(&w1, nHiddenNodes, nInputNodes + 1);
        MatrixAllocate(&target, nPatterns, nOutputNodes);
        VectorAllocate(&PatternID, nPatterns);

        //读初始权矩阵 
        if ((fpWeights = fopen(szWeights, "r")) == NULL)
        {
            fprintf(stderr, "%s:can't open file %s\n", progname, szWeights);
            system("pause");
            exit(1);
        }

        //读输入-隐含层权值 
        for (h = 0; h < nHiddenNodes; h++)
            for (i = 0; i <= nInputNodes; i++)
            {
                fscanf(fpWeights, "%lf", &w1[h][i]); // 修复：%f改为%lf
                delw1[h][i] = 0.0;
            }

        //读隐含-输出层的权值 
        for (j = 0; j < nOutputNodes; j++)
            for (h = 0; h <= nHiddenNodes; h++)
            {
                fscanf(fpWeights, "%lf", &w2[j][h]); // 修复：%f改为%lf
                delw2[j][h] = 0.0; // 修复：原代码误写为delw1，且越界，改为delw2防止崩溃
            }

        //读所有要学习的模式 
        if ((fpPattern = fopen(szPattern, "r")) == NULL)
        {
            fprintf(stderr, "%s:can't open file %s\n", progname, szPattern);
            system("pause");
            exit(1);
        }

        for (p = 0; p < nPatterns; p++)
        {
            for (i = 0; i < nInputNodes; i++)
                if (fscanf(fpPattern, "%lf", &out0[p][i]) != 1) // 修复：原代码写成了"&f"，且double需用%lf
                    goto ALLPATTERNSREAD;

            //读目标输出 
            for (j = 0; j < nOutputNodes; j++) // 修复：原代码为j<=nOutputNodes，会导致target数组严重越界崩溃
                fscanf(fpPattern, "%lf", &target[p][j]); // 修复：%f改为%lf

            //读每一个模式的分别号 
            fscanf(fpPattern, "%lf", &PatternID[p]); // 修复：%f改为%lf
        }

    ALLPATTERNSREAD:
        fclose(fpPattern);

        if (p < nPatterns)
        {
            fprintf(stderr, "%s:%d out of %d patterns read\n", progname, p, nPatterns);
            nPatterns = p;
        }

        //打开误差输出文件 
        if ((fpError = fopen(szError, "w")) == NULL)
        {
            fprintf(stderr, "%s:can't open file %s\n", progname, szError);
            system("pause");
            exit(1);
        }

        fprintf(stderr, nIterations > 1 ? "Training...\n" : "Testing\n"); // 修复：原代码拼写错误Teating

        //开始迭代循环 
        for (q = 0; q < nIterations; q++)
        {
            for (p = 0; p < nPatterns; p++)
            {
                //隐含层 
                for (h = 0; h < nHiddenNodes; h++)
                {
                    double sum = w1[h][nInputNodes];
                    for (i = 0; i < nInputNodes; i++)
                        sum += w1[h][i] * out0[p][i];
                    //计算输出层 
                    out1[p][h] = 1.0 / (1.0 + exp(-sum));
                }

                //输出层 
                for (j = 0; j < nOutputNodes; j++)
                {
                    double sum = w2[j][nHiddenNodes];
                    for (h = 0; h < nHiddenNodes; h++)
                        sum += w2[j][h] * out1[p][h];
                    out2[p][j] = 1.0 / (1.0 + exp(-sum));
                }

                //输出层误差修正量 
                for (j = 0; j < nOutputNodes; j++)
                    delta2[p][j] = (target[p][j] - out2[p][j]) * out2[p][j] * (1.0 - out2[p][j]);

                //隐含层误差修正量 
                for (h = 0; h < nHiddenNodes; h++) // 修复：原代码错写为h<nOutputNodes
                {
                    double sum = 0.0;
                    for (j = 0; j < nOutputNodes; j++)
                        sum += delta2[p][j] * w2[j][h];
                    delta1[p][h] = sum * out1[p][h] * (1.0 - out1[p][h]);
                }
            }

            //调整隐含-输出层部的权值 
            for (j = 0; j < nOutputNodes; j++)
            {
                double dw;
                double sum = 0.0;
                for (p = 0; p < nPatterns; p++)
                    sum += delta2[p][j];
                dw = eta * sum + alpha * delw2[j][nHiddenNodes];
                w2[j][nHiddenNodes] += dw;
                delw2[j][nHiddenNodes] = dw;

                for (h = 0; h < nHiddenNodes; h++)
                {
                    double sum = 0.0;
                    for (p = 0; p < nPatterns; p++)
                        sum += delta2[p][j] * out1[p][h];
                    dw = eta * sum + alpha * delw2[j][h];
                    w2[j][h] += dw;
                    delw2[j][h] = dw;
                }
            }

            //调整输入-隐含层的权值 
            for (h = 0; h < nHiddenNodes; h++)
            {
                double dw;
                double sum = 0.0;
                for (p = 0; p < nPatterns; p++)
                    sum += delta1[p][h];
                dw = eta * sum + alpha * delw1[h][nInputNodes]; // 修复：原代码误用j导致随机值越界崩溃
                w1[h][nInputNodes] += dw;
                delw1[h][nInputNodes] = dw; // 修复：同上

                for (i = 0; i < nInputNodes; i++)
                {
                    double sum = 0.0;
                    for (p = 0; p < nPatterns; p++)
                        sum += delta1[p][h] * out0[p][i];
                    dw = eta * sum + alpha * delw1[h][i];
                    w1[h][i] += dw;
                    delw1[h][i] = dw;
                }
            }

            //监督键盘响应 
            if (_kbhit())
            {
                int c = _getch(); // 修复：VS标准推荐使用_getch
                if (toupper(c) == 'E') // 修复：原代码 if(c=toupper(c)=='E') 逻辑错误，会变成赋值1或0
                    MonitorError++;
                else if (c == ESC)
                    break;
            }

            //计算均方根误差 
            if (MonitorError || (q % nReportErrors == 0))
            {
                for (p = 0, error = 0.0; p < nPatterns; p++)
                {
                    for (j = 0; j < nOutputNodes; j++)
                    {
                        double temp = target[p][j] - out2[p][j];
                        error += temp * temp;
                    }
                }
                error /= (nPatterns * nOutputNodes);
                fprintf(stderr, "iteration %d-%d Error %f\r", q, nIterations, error); // 修复：原格式符"%5d%-5d"错误
                MonitorError = 0;

                if (q % nReportErrors == 0)
                    fprintf(fpError, "%d %f\n", q, error);

                if (error < ErrorLevel)
                    break;
            }
        }

        //迭代循环结束 
        for (p = 0, error = 0.0; p < nPatterns; p++)
        {
            for (j = 0; j < nOutputNodes; j++)
            {
                double temp = target[p][j] - out2[p][j];
                error += temp * temp;
            }
        }

        //计算所有模式的平均误差 
        error /= (nPatterns * nOutputNodes);
        fprintf(stderr, "Iteration %d-%d Error %f\n", q, nIterations, error); // 修复：格式符错误
        fprintf(fpError, "\n%d %f\n", q, error);
        fclose(fpError);

        if ((fpWeightsOut = fopen(szWeightsOut, "w")) == NULL)
        {
            fprintf(stderr, "%s:can't write file %s\n", progname, szWeightsOut);
            system("pause");
            exit(1);
        }

        for (h = 0; h < nHiddenNodes; h++)
            for (i = 0; i <= nInputNodes; i++)
                fprintf(fpWeightsOut, "%g%c", w1[h][i], i % ITEMS == ITEMS - 1 ? '\n' : ' ');

        for (j = 0; j < nOutputNodes; j++)
            for (h = 0; h < nHiddenNodes; h++)
                fprintf(fpWeightsOut, "%g%c", w2[j][h], j % ITEMS == ITEMS - 1 ? '\n' : ' ');

        fclose(fpWeightsOut);

        if ((fpResults = fopen(szResults, "w")) == NULL)
        {
            fprintf(stderr, "%s:can't write file %s\n", progname, szResults);
            fpResults = stderr;
        }

        for (p = 0; p < nPatterns; p++)
        {
            fprintf(fpResults, "%d ", p);
            for (j = 0; j < nOutputNodes; j++)
                fprintf(fpResults, "%f ", out2[p][j]); // 修复：补充空格分隔
            fprintf(fpResults, " %-6.0f\n", PatternID[p]);
        }

        fclose(fpResults);

        //释放数据的动态存储 
        MatrixFree(out0, nPatterns);
        MatrixFree(out1, nPatterns);
        MatrixFree(delta1, nPatterns);
        MatrixFree(delw1, nHiddenNodes);
        MatrixFree(w1, nHiddenNodes);
        MatrixFree(out2, nPatterns);
        MatrixFree(delta2, nPatterns);
        MatrixFree(delw2, nOutputNodes);
        MatrixFree(w2, nOutputNodes);
        MatrixFree(target, nPatterns);
        free(PatternID);
    }

    fclose(fpRun);
    system("pause"); // 防止VS调试结束时闪退
    return 0; // 配合int main
}

void VectorAllocate(VECTOR* vector, int nCols)
{
    if ((*vector = (VECTOR)calloc(nCols, sizeof(double))) == NULL)
    {
        fprintf(stderr, "Sorry! Not enough memory for nodes\n");
        system("pause");
        exit(1);
    }
}

void AllocateCols(PFLOAT matrix[], int nRows, int nCols)
{
    int i;
    for (i = 0; i < nRows; i++)
        VectorAllocate(&matrix[i], nCols);
}

void MatrixAllocate(MATRIX* pmatrix, int nRows, int nCols)
{
    if ((*pmatrix = (MATRIX)calloc(nRows, sizeof(PFLOAT))) == NULL)
    {
        fprintf(stderr, "Sorry! Not enough memory for nodes\n");
        system("pause");
        exit(1);
    }
    AllocateCols(*pmatrix, nRows, nCols);
}

void MatrixFree(MATRIX matrix, int nRows)
{
    int i;
    for (i = 0; i < nRows; i++)
        free(matrix[i]);
}
